internal/pkg/skuba/addons/cilium.go

```diff
@@ -585,5 +639,572 @@ rules:
+          valueFrom:
+            fieldRef:
+              apiVersion: v1
+              fieldPath: spec.nodeName
+        - name: CILIUM_K8S_NAMESPACE
+          valueFrom:
+            fieldRef:
+              apiVersion: v1
+              fieldPath: metadata.namespace
+        - name: CILIUM_FLANNEL_MASTER_DEVICE
+          valueFrom:
+            configMapKeyRef:
+              key: flannel-master-device
+              name: cilium-config
+              optional: true
+        - name: CILIUM_FLANNEL_UNINSTALL_ON_EXIT
+          valueFrom:
+            configMapKeyRef:
+              key: flannel-uninstall-on-exit
+              name: cilium-config
+              optional: true
+        - name: CILIUM_CLUSTERMESH_CONFIG
+          value: /var/lib/cilium/clustermesh/
+        - name: CILIUM_CNI_CHAINING_MODE
+          valueFrom:
+            configMapKeyRef:
+              key: cni-chaining-mode
+              name: cilium-config
+              optional: true
+        - name: CILIUM_CUSTOM_CNI_CONF
+          valueFrom:
+            configMapKeyRef:
+              key: custom-cni-conf
+              name: cilium-config
+              optional: true
+        image: {{.CiliumImage}}
+        imagePullPolicy: IfNotPresent
+        lifecycle:
+          postStart:
+            exec:
+              command:
+              - cilium-cni-install
+          preStop:
+            exec:
+              command:
+              - cilium-cni-uninstall
+        livenessProbe:
+          exec:
+            command:
+            - cilium
+            - status
+            - --brief
+          failureThreshold: 10
+          # The initial delay for the liveness probe is intentionally large to
+          # avoid an endless kill & restart cycle if in the event that the initial
+          # bootstrapping takes longer than expected.
+          initialDelaySeconds: 120
+          periodSeconds: 30
+          successThreshold: 1
+          timeoutSeconds: 5
+        name: cilium-agent
+        readinessProbe:
+          exec:
+            command:
+            - cilium
+            - status
+            - --brief
+          failureThreshold: 3
+          initialDelaySeconds: 5
+          periodSeconds: 30
+          successThreshold: 1
+          timeoutSeconds: 5
+        securityContext:
+          capabilities:
+            add:
+            - NET_ADMIN
+            - SYS_MODULE
+          privileged: true
+        volumeMounts:
+        - mountPath: /var/run/cilium
+          name: cilium-run
+        - mountPath: /host/opt/cni/bin
+          name: cni-path
+        - mountPath: /host/etc/cni/net.d
+          name: etc-cni-netd
+        - mountPath: /var/run/crio/crio.sock
+          name: crio-socket
+{{if .AddonContext.NeedsMigration}}
+        - mountPath: /var/lib/etcd-config
+          name: etcd-config-path
+          readOnly: true
+        - mountPath: /var/lib/etcd-secrets
+          name: etcd-secrets
+          readOnly: true
+{{end}}
```

Comment:
I'm thinking if we could do this without adding a new AddonContext callback.

But generate a base manifest, and add patch manifest for this as [kustomize](https://github.com/kubernetes-sigs/kustomize) does.

Reply:
That's a great idea. I will need to modify a bit the way we are using kustomize now, to be able to use particular patches for different cases (i.e. use the patch with etcd config only if we need to perform the migration). But overall it will be much better from what I did here. thanks!

Reply:
From what I see now, currently skuba allows users to provide patches in `patches/` directory. Defining our own patches and rendering them would be something new to do. I think we should define a list/map of patches in the `Addon` struct, provide some way to enable/disable them according to circumstances, then copy them to some other directory like `skuba-patches/`, to differentiate them from patches provides by users.

@ereslibre WDYT?

Reply:
I would keep the current approach. It's simpler than modifying the template. I don't see a big value added by making the patch a kustomize patch, as we'll need to remove things anyway later.

Reply:
I will actually use even different approach, the third way. ;)

I realized that a good way to perform a migration is not to use the main Cilium daemonset, but rather to make a very dummy preflight deployment which just runs `sleep` on Cilium images and reads the etcd config. I will deploy that preflight daemonset as a step in `beforeApply`.

Reply:
That's very clean.

Reply:
So far the preflight ds solution seems to work.